tools-local/tasks/Microsoft.DotNet.SourceBuild.Tasks/CopyReferenceOnlyPackages.cs

```diff
@@ -0,0 +1,116 @@
+    {
+        /// <summary>
+        /// Package cache dir containing prebuilt nupkgs. Path is expected to be like:
+        /// 
+        /// {PackageCacheDir}/{lowercase id}/{version}/{lowercase id}.{version}.nupkg
+        ///
+        /// This code assumes that these are expanded packages loaded from nuget.
+        /// </summary>
+        [Required]
+        public string PackageCacheDir { get; set; }
+
+        /// <summary>
+        /// The destination directory for the dlls in reference packages.
+        /// Paths are preserved in the destination.
+        /// </summary>
+        [Required]
+        public string DllDestinationDir { get; set; }
+
+        /// <summary>
+        /// The destination directory for the nupkg files identified 
+        /// as containing reference-only packages.
+        /// </summary>
+        [Required]
+        public string IdentifiedPackagesDir { get; set; }
+
+        /// <summary>
+        /// The destination directory for the reference packages.
+        /// </summary>
+        [Required]
+        public string DestinationDir { get; set; }
+
+        public override bool Execute()
+        {
+            DateTime startTime = DateTime.Now;
+
+            var referenceOnlyPackageDirectories = Directory.EnumerateFiles(PackageCacheDir, "*.nupkg", SearchOption.AllDirectories)
+                .Where(nupkgFilePath =>
+                {
+                    // Get all files in the nupkg path and normalize directory separators
+                    var nupkgFiles = Directory.EnumerateFiles(Path.GetDirectoryName(nupkgFilePath), "*.*", SearchOption.AllDirectories).Select(f => f.Replace("\\","/"));
+
+                    // Do not include directories that contain exes, shared object files, OSX dynamic libraries
+                    // or profiling data
+                    string[] extensionsToExclude = { ".exe", ".dylib", ".so", ".profdata", ".pgd" };
+                    string[] pathsToExclude = { "testdata" };
+                    
+                    // Return directories that, if containing dlls, only have dlls in the
+                    // ref folder.
+                    return
+                        !nupkgFiles.Any(file => extensionsToExclude.Contains(Path.GetExtension(file)) || pathsToExclude.Any(path => file.Contains(path))) &&
+                        nupkgFiles
+                            .Where(file => String.Equals(Path.GetExtension(file), ".dll", StringComparison.OrdinalIgnoreCase))
+                            .All(dir => dir.Contains(@"/ref/"));
+                })
+                .Select(f => Path.GetDirectoryName(f));
+
+            Directory.CreateDirectory(IdentifiedPackagesDir);
+            foreach (var dir in referenceOnlyPackageDirectories)
+            {
+                Directory.CreateDirectory(dir.Replace(PackageCacheDir, DestinationDir));
+                foreach (string dirPath in Directory.EnumerateDirectories(dir, "*", SearchOption.AllDirectories))
```

Comment:
Could you save the enumeration results used to determine that this is a ref-only package and use it here? (Keep it in an anonymous type?)